### PR TITLE
ports api

### LINKED ISSUE #49: command not found: opened-ports
https://github.com/canonical/operator/pull/905 introduced new commands that scenario does not yet support.

```
Traceback (most recent call last):
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/scenario/runtime.py", line 387, in exec
    mocked_main(
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/scenario/ops_main_mock.py", line 132, in main
    charm = charm_class(framework)
  File "/home/code/o11y/alertmanager-k8s-operator/src/charm.py", line 131, in __init__
    self.set_ports()
  File "/home/code/o11y/alertmanager-k8s-operator/src/charm.py", line 232, in set_ports
    actual_ports = self.unit.opened_ports()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/ops/model.py", line 630, in opened_ports
    return self._backend.opened_ports()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/ops/model.py", line 3154, in opened_ports
    output = typing.cast(str, self._run('opened-ports', return_output=True))
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/ops/model.py", line 2686, in _run
    raise RuntimeError(f'command not found: {args[0]}')
RuntimeError: command not found: opened-ports

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/runner.py", line 341, in from_call
    result: Optional[TResult] = func()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_callers.py", line 155, in _multicall
    return outcome.get_result()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/runner.py", line 157, in pytest_runtest_setup
    item.session._setupstate.setup(item)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/runner.py", line 497, in setup
    raise exc
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/runner.py", line 494, in setup
    col.setup()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/python.py", line 1791, in setup
    self._request._fillfixtures()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 566, in _fillfixtures
    item.funcargs[argname] = self.getfixturevalue(argname)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 585, in getfixturevalue
    fixturedef = self._get_active_fixturedef(argname)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 607, in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 693, in _compute_fixture_value
    fixturedef.execute(request=subrequest)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 1069, in execute
    result = ihook.pytest_fixture_setup(fixturedef=self, request=request)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_callers.py", line 155, in _multicall
    return outcome.get_result()
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_result.py", line 108, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 1123, in pytest_fixture_setup
    result = call_fixture_func(fixturefunc, request, kwargs)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/_pytest/fixtures.py", line 895, in call_fixture_func
    fixture_result = next(generator)
  File "/home/code/o11y/alertmanager-k8s-operator/tests/scenario/test_server_scheme.py", line 35, in initial_state
    state = begin_with_initial_hooks_isolated(context, leader=leader)
  File "/home/code/o11y/alertmanager-k8s-operator/tests/scenario/helpers.py", line 12, in begin_with_initial_hooks_isolated
    state = context.run("install", state)
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/scenario/context.py", line 151, in run
    return self._run(
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/scenario/context.py", line 221, in _run
    return runtime.exec(
  File "/home/code/o11y/alertmanager-k8s-operator/.tox/scenario/lib/python3.10/site-packages/scenario/runtime.py", line 400, in exec
    raise UncaughtCharmError(
scenario.runtime.UncaughtCharmError: Uncaught exception (<class 'RuntimeError'>) in operator/charm code: RuntimeError('command not found: opened-ports')
```

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds State and mocking api for the recently-added-to-ops `[open/closed/opened]-ports` wrappers.

Since `ops 2.6.0`, charms can invoke the `open-port`, `close-port`, and `opened-ports` hook tools to manage the ports opened on the host vm/container. Using the `State.opened_ports` api, you can: 

- simulate a charm run with a port opened by some previous execution
```python
from scenario import State, Port, Context

ctx = Context(MyCharm)
ctx.run("start", State(opened_ports=[Port("tcp", 42)]))
```
- assert that a charm has called `open-port` or `close-port`:
```python
from scenario import State, Port, Context

ctx = Context(MyCharm)
state1 = ctx.run("start", State())
assert state1.opened_ports == [Port("tcp", 42)]

state2 = ctx.run("stop", state1)
assert state2.opened_ports == []
```

Fixes #49

### CODE REVIEW

**Comment by benhoyt** [comment]
Just wanted to note the `open-port` (etc) support has been in place since `2.1.0` (see [release notes](https://github.com/canonical/operator/releases/tag/2.1.0)). I'd suggest not bumping the ops requirement (you could even leave it on >=2.0 -- if folks don't need this, they can still use ops 2.0.0 if they really want).

File: pyproject.toml [hunk 1]
```diff
@@ -8,7 +8,7 @@ build-backend = "setuptools.build_meta"
 [project]
 name = "ops-scenario"
 
-version = "5.0"
+version = "5.0.1"
```
Comment (benhoyt) [inline pyproject.toml:11]: This adds a whole new feature, right, so it would be a minor version bump, 5.1.0?

**Review by benhoyt — APPROVED** [review]

File: scenario/state.py [hunk 2]
```diff
@@ -755,6 +759,24 @@ def handle_path(self):
...
+    def __post_init__(self):
+        port = self.port
+        if self.protocol == "icmp" and port:
+            raise StateValidationError("`port` arg not supported with `icmp` protocol")
+        elif not port:
+            raise StateValidationError(
+                f"`port` arg required with `{self.protocol}` protocol",
+            )
+        if port and not (1 <= port <= 65535):
+            raise StateValidationError(f"`port` outside bounds [1:65535], got {port}")
```
Comment (Abuelodelanada) [inline scenario/state.py:777]: This is obviously a matter of taste, but the use of clause guards gives us a slightly cleaner code.


```suggestion
        if not port:
            raise StateValidationError(
                f"`port` arg required with `{self.protocol}` protocol",
            )
        if self.protocol == "icmp":
            raise StateValidationError("`port` arg not supported with `icmp` protocol")

        if not (1 <= port <= 65535):
            raise StateValidationError(f"`port` outside bounds [1:65535], got {port}")
```

**Review by Abuelodelanada — APPROVED** [review]

File: scenario/state.py (hunk 2 above)
Comment (PietroPasotti) [reply scenario/state.py:777]: you're right, it deserved a cleanup, but I ended up doing it slightly differently.

File: pyproject.toml (hunk 1 above)
Comment (PietroPasotti) [reply pyproject.toml:11]: 5.1.0 it is

**Pull request merged**